sematic/db/models/edge.py

```diff
@@ -19,20 +19,20 @@ class Edge(Base, JSONEncodableMixin):
 
     # Edge endpoints
     source_run_id: Optional[str] = Column(
-        types.String(), ForeignKey("artifacts.id"), nullable=True
+        types.String(), ForeignKey("runs.id"), nullable=True
     )
     source_name: Optional[str] = Column(types.String(), nullable=True)
     destination_run_id: Optional[str] = Column(
-        types.String(), ForeignKey("artifacts.id"), nullable=True
+        types.String(), ForeignKey("runs.id"), nullable=True
     )
     destination_name: Optional[str] = Column(types.String(), nullable=True)
 
     # Artifact
     artifact_id: Optional[str] = Column(
-        types.String(), ForeignKey("artifacts.id"), nullable=False
+        types.String(), ForeignKey("artifacts.id"), nullable=True
```

Comment:
turns out that this should actually be nullable, because edges are created before the result of the run is known, and failed runs will not have artifacts, so there are many instances in the db where `artifact_id` is null.